src/lib/common/sensorWatchdog.cpp

```diff
@@ -0,0 +1,131 @@
+    uint32_t _triggerCount;
+} sensor_watchdog_t;
+
+static sensor_watchdog_t * _sensor_watchdog_list = NULL;
+static SemaphoreHandle_t _sensor_watchdog_list_mutex = NULL;
+
+static void _addSensorWatchdog(sensor_watchdog_t * cfg);
+static void _timerCallback(TimerHandle_t xTimer);
+static void _sensorWatchdogHandler(void *arg);
+
+/*!
+ * @brief Initialize the sensor watchdog.
+ */
+void SensorWatchdogInit(void) {
+    _sensor_watchdog_list_mutex = xSemaphoreCreateMutex();
+}
+
+/*!
+ * @brief Add a sensor watchdog.
+ * @param[in] id The id of the sensor watchdog.
+ * @param[in] timeoutMs The timeout in milliseconds.
+ * @param[in] handler The callback function to be called when the sensor watchdog expires.
+ * @param[in] max_triggers The maximum number of times the sensor watchdog can expire before it is stopped.
+ * @param[in] logHandle The spotter BM log handle to log to. If NULL, no logging will be done.
+ */
+void SensorWatchdogAdd(const char * id, uint32_t timeoutMs, sensor_watchdog_handler handler, uint32_t max_triggers, const char * logHandle) {   
+    configASSERT(handler);
+    configASSERT(timeoutMs > 0);
+    configASSERT(id);
+    sensor_watchdog_t * cfg = static_cast<sensor_watchdog_t *>(pvPortMalloc(sizeof(sensor_watchdog_t)));
+    configASSERT(cfg);
+    cfg->_id = id;
+    cfg->_timeoutMs = timeoutMs;
+    cfg->_handler = handler;
+    cfg->_logHandle = logHandle;
+    cfg->_max_triggers = max_triggers;
+    cfg->_triggerCount = 0;
+    cfg->_timerHandle =  xTimerCreate(cfg->_id, pdMS_TO_TICKS(cfg->_timeoutMs), pdTRUE, NULL, _timerCallback);
+    cfg->_next = NULL;
+    _addSensorWatchdog(cfg);
+    configASSERT(xTimerStart(cfg->_timerHandle, 0) == pdTRUE);
+}
+
+/*!
+ * @brief Refresh a sensor watchdog.
+ * @param[in] id The id of the sensor watchdog to pet.
+ */
+void SensorWatchdogPet(const char * id) {
+    configASSERT(id);
+    configASSERT(xSemaphoreTake(_sensor_watchdog_list_mutex, pdMS_TO_TICKS(WATCHDOG_PET_TIMEOUT_MS)) == pdTRUE);
+    sensor_watchdog_t * curr = _sensor_watchdog_list;
+    while(curr) {
+        if(strcmp(curr->_id, id) == 0) {
+            curr->_triggerCount = 0;
+            configASSERT(xTimerReset(curr->_timerHandle, pdMS_TO_TICKS(WATCHDOG_PET_TIMEOUT_MS)) == pdTRUE);
+            break;
+        }
+        curr = curr->_next;
+    }
+    xSemaphoreGive(_sensor_watchdog_list_mutex);
+}
+
+static void _addSensorWatchdog(sensor_watchdog_t * cfg) {
+    configASSERT(cfg);
+    xSemaphoreTake(_sensor_watchdog_list_mutex, portMAX_DELAY);
+    if (_sensor_watchdog_list) {
+        sensor_watchdog_t * curr = _sensor_watchdog_list;
+        while (curr->_next) {
+            curr = curr->_next;
+        }
+        curr->_next = cfg;
+    } else {
+        _sensor_watchdog_list = cfg;
+    }
+    xSemaphoreGive(_sensor_watchdog_list_mutex);
+}
+
+static void _timerCallback(TimerHandle_t xTimer) {
+    configASSERT(xTimer);
+    configASSERT(xSemaphoreTake(_sensor_watchdog_list_mutex, 0) == pdTRUE);
```

Comment:
We block for zero in `_timerCallback` because:
```
Timer callback functions execute in the context of the timer service task. It is therefore essential that timer callback functions never attempt to block. For example, a timer callback function must not call vTaskDelay(), vTaskDelayUntil(), or specify a non zero block time when accessing a queue or a semaphore.
```
And instead we attempt to run in the callback task: `timer_handler`